src/dust.ts

```diff
@@ -177,6 +177,36 @@ export class Dust {
         return state;
     }
 
+    /**
+     * Reorder particle state among particles
+     *
+     * @param index An integer array of length `nParticles`, with
+     * values between 0 and `nParticles`, indicating the index of the
+     * particle that each particle should end up taking. So if the
+     * `i`th element is `j`, then after reordering particle `i` will
+     * have the state that particle `j` currently has. Indices can be
+     * repeated, and typically will be; i.e., this is a resampling
+     * rather than a strict shuffle.
```

Comment:
OK, so you could end up with multiple 'copies' of some particles, while removing some altogether (if the number of particles must stay the same)? Is `reorder` the best name for the function? Would `resample` be better?

Reply:
Resample (IMO) implies the presence of the sampling process. We do have a resample method on the C++ version of this which does a sample, then reorders with that index: https://github.com/mrc-ide/dust/blob/master/inst/include/dust/dust_cpu.hpp#L234-L263

Because we have a bunch more flexibility with JS than with C++ it's likely we'll implement the filter outside of this class so we may never end up with a resample method here, but it would still be good to keep the names somewhat aligned